activestorage/app/controllers/active_storage/public_disk_controller.rb

```diff
@@ -7,14 +7,23 @@ class ActiveStorage::PublicDiskController < ActiveStorage::BaseController
   skip_forgery_protection
 
   def show
-    if blob = ActiveStorage::Blob.find_by(key: params[:key])
+    if blob = ActiveStorage::Blob.find_by(key: extract_blob_key)
       if blob.service.public?
-        serve_file blob.service.path_for(blob.key), content_type: blob.content_type, disposition: :inline
+        serve_file blob.service.path_for(params[:key]), content_type: blob.content_type, disposition: :inline
       else
         head :unauthorized
       end
     else
       head :not_found
     end
   end
+
+  private
+    def extract_blob_key
```

Comment:
I don't know if the variant key is meant to be plugged into a URL parameter unsigned, hence why we are encountering this issue. 

Maybe we could change how URLs are generated and make `ActiveStorage::Variation#key` a `variation` query parameter to load variants. Otherwise `variant/{original_blob_key}` would be valid, which doesn't make sense.

Reply:
I'm not sure if I understand your second paragraph, could you iterate on it a little more? Are you suggesting to change the format of the string returned by `ActiveStorage::Variant#key`, or update the routes to allow a `/disk/public/variant/:key/:variant/*filename` route?

Reply:
Just the route, and we could use `ActiveStorage::Variation#key` as the distinguishing query parameter to tell the controller to load the variant (and not the base blob) eg. `/disk/public/:key/*filename?variation=variation_key_here`.

However, we could also try to fix this by making the public disk controller more like the private one and forgo loading the blob and use an `encoded_key` instead, the public one just wouldn't have an expiry set. This also hides the backing key from the client.

Reply:
I think I found a better solution in #37897. Please take a look it!